client/src/plugins/usage-statistics/__tests__/UsageStatisticsSpec.js

```diff
@@ -0,0 +1,397 @@
+    });
+
+    const enableSpy = sinon.spy();
+
+    instance.enable = enableSpy;
+
+    // when
+    await instance.componentDidMount();
+
+    // then
+    expect(enableSpy).to.not.have.been.called;
+  });
+
+
+  it('should enable when mounted', async () => {
+
+    // given
+    Flags.init({
+      [ ET_ENDPOINT ]: 'test-et-endpoint'
+    });
+
+    const { instance } = createUsageStatistics({
+      configValues: { 'editor.privacyPreferences': { ENABLE_USAGE_STATISTICS: true } }
+    });
+
+    const enableSpy = sinon.spy();
+
+    instance.enable = enableSpy;
+
+    // when
+    await instance.componentDidMount();
+
+    // then
+    expect(enableSpy).to.have.been.called;
+  });
+
+
+  it('should enable registered event handlers', async () => {
+
+    // given
+    Flags.init({
+      [ ET_ENDPOINT ]: 'test-et-endpoint'
+    });
+
+    const { instance } = createUsageStatistics({
+      configValues: { 'editor.privacyPreferences': { ENABLE_USAGE_STATISTICS: true } }
+    });
+
+    const eventHandler = new BaseEventHandler('test-event', () => {});
+
+    instance._eventHandlers = [ eventHandler ];
+
+    // when
+    await instance.componentDidMount();
+
+    // then
+    expect(eventHandler.isEnabled()).to.eql(true);
+  });
+
+
+  it('should disable registered event handlers', async () => {
+
+    // given
+    Flags.init({
+      [ ET_ENDPOINT ]: 'test-et-endpoint'
+    });
+
+    const { instance } = createUsageStatistics({
+      configValues: { 'editor.privacyPreferences': { ENABLE_USAGE_STATISTICS: true } }
+    });
+
+    const eventHandler = new BaseEventHandler('test-event', () => {});
+
+    instance._eventHandlers = [ eventHandler ];
+
+    // when
+    await instance.componentDidMount();
+
+    instance.disable();
+
+    // then
+    expect(eventHandler.isEnabled()).to.eql(false);
+  });
+
+
+  it('should listen to privacy preference change event', async () => {
+
+    // given
+    Flags.init({
+      [ ET_ENDPOINT ]: 'test-et-endpoint'
+    });
+
+    const subscribeSpy = sinon.spy();
+
+    const { instance } = createUsageStatistics({
+      configValues: { 'editor.privacyPreferences': { ENABLE_USAGE_STATISTICS: true } },
+      subscribeSpy
+    });
+
+    instance.fetch = () => {};
+
+    // when
+    await instance.componentDidMount();
+
+    // then
+    expect(subscribeSpy).to.have.been.calledWith('privacy-preferences.changed');
+  });
+
+
+  it('should enable when usage statistics preference switched on', async () => {
+
+    // given
+    Flags.init({
+      [ ET_ENDPOINT ]: 'test-et-endpoint'
+    });
+
+    const subscribeSpy = sinon.spy();
+
+    let usageStatisticsEnabled = false;
+
+    const { instance } = createUsageStatistics({
+      configGet: () => {
+        return { ENABLE_USAGE_STATISTICS: usageStatisticsEnabled };
+      },
+      subscribeSpy
+    });
+
+    const enableSpy = sinon.spy();
+    instance.enable = enableSpy;
+
+    // when
+    await instance.componentDidMount();
```

Comment:
This calls the method for the second time, as you can test in a simple test case:

```javascript
it.only('componentDidMount', async function() {

  // given
  const spy = sinon.spy();
  class Component extends PureComponent {
    async componentDidMount() {
      spy();
    }

    render() {
      return null;
    }
  }

  // when
  const wrapper = shallow(<Component />);

  await (wrapper.instance()).componentDidMount();

  // then
  expect(spy).to.have.property('calledTwice', true);
});
```

Generally, I'd rather avoid calling internal methods within test cases. `shallow` already does it.

Reply:
I'm aware this calls the componentDidMount twice. However this behavior do not affect our tests. For the tests, we usually create the component (at this point componentDidMount is already performed), then we set some dummy stuff which are used inside the `componentDidMount` logic depending on our tests, then we re-mount the component to perform our tests.

We have a similar approach in [ErrorTracking tests](https://github.com/camunda/camunda-modeler/blob/develop/client/src/plugins/error-tracking/__tests__/ErrorTrackingSpec.js) as well. This plugin and Error tracking plugin are non-visual plugins (they do not render anything), that's why most of the things get done within the `componentDidMount`.

Reply:
The question is whether we test real component behavior when we call `componentDidMount` manually. In the current state, as a side effect, logging happens twice and the plugin might be enabled twice without disabling in between.

Reply:
Your point makes sense. I'll investigate this.

Another approach would be -> Since this is a no-render component, maybe we can directly test it without using enzyme? We don't need DOM queries etc.

Reply:
why not

Reply:
Here we go: https://github.com/camunda/camunda-modeler/pull/1840/commits/15ca8ca58ec8b864d2bbc31e77afe0842d66f365

I like this solution the best. We don't really need enzyme there. I'll do the same for ErrorTracking tests with a different PR after we merge this.

Reply:
Yeah, but please note that at least one render test should still remain. We still do render the component in the app after all.

Reply:
Ok, I brought back the `should render` test: https://github.com/camunda/camunda-modeler/commit/78486063049b99a56d1769ae0811198da6a4645f